src/lib.rs

```diff
@@ -14,75 +21,88 @@ pub trait OwnedCharsExt {
 
 impl OwnedCharsExt for String {
     fn into_chars(self) -> OwnedChars {
-        OwnedChars { s: self, i: 0 }
+        OwnedChars::from_string(self)
     }
 
     fn into_char_indices(self) -> OwnedCharIndices {
-        OwnedCharIndices { s: self, i: 0 }
+        OwnedCharIndices::from_string(self)
     }
 }
 
 /// Iterator over the chars of a string (the string is owned by the iterator)
-#[derive(Clone, Debug)]
+#[derive(Debug)]
 pub struct OwnedChars {
     s: String,
-    i: usize,
+    i: Chars<'static>,
 }
 
 /// Iterator over the chars of a string and their indices (the string is owned by the iterator)
-#[derive(Clone, Debug)]
+#[derive(Debug)]
 pub struct OwnedCharIndices {
     s: String,
-    i: usize,
+    i: CharIndices<'static>,
 }
 
 macro_rules! impls {
-    ($s:ident) => {
-        impl $s {
+    ($struct:ident, $method: ident, $item: ty) => {
+        impl $struct {
+            /// Create Self from a String, moving the String into Self
+            fn from_string(s: String) -> Self {
+                // First, move the string
+                let mut owned = $struct {
+                    s: s,
+                    i: unsafe { uninitialized() }
```

Comment:
My personal "unsafe code policy" is that when leaving an `unsafe` block, the invariants that we are screwing with should be restored. So could you just merge these two blocks so that lines 52-60 are under one `unsafe`?